Source/Core/VideoCommon/UberShaderVertex.cpp

```diff
@@ -50,15 +54,90 @@ ShaderCode GenVertexShader(APIType api_type, const ShaderHostConfig& host_config
   WriteBitfieldExtractHeader(out, api_type, host_config);
   WriteLightingFunction(out);
 
-  out.Write("ATTRIBUTE_LOCATION({}) in float4 rawpos;\n", SHADER_POSITION_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in uint4 posmtx;\n", SHADER_POSMTX_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in float3 rawnormal;\n", SHADER_NORMAL_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in float3 rawtangent;\n", SHADER_TANGENT_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in float3 rawbinormal;\n", SHADER_BINORMAL_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in float4 rawcolor0;\n", SHADER_COLOR0_ATTRIB);
-  out.Write("ATTRIBUTE_LOCATION({}) in float4 rawcolor1;\n", SHADER_COLOR1_ATTRIB);
-  for (int i = 0; i < 8; ++i)
-    out.Write("ATTRIBUTE_LOCATION({}) in float3 rawtex{};\n", SHADER_TEXTURE0_ATTRIB + i, i);
+  if (host_config.backend_dynamic_vertex_loader)
+  {
+    out.Write(R"(
+SSBO_BINDING(1) readonly restrict buffer Vertices {{
+  uint vertex_buffer[];
+}};
+)");
+    if (api_type == APIType::D3D)
+    {
+      // D3D doesn't include the base vertex in SV_VertexID
+      out.Write("UBO_BINDING(std140, 3) uniform DX_Constants {{\n"
+                "  uint base_vertex;\n"
+                "}};\n\n"
+                "uint GetOffset(uint attr_offset) {{\n"
+                "  return (gl_VertexID + base_vertex) * vertex_stride + attr_offset;\n"
+                "}}\n");
+    }
+    else
+    {
+      out.Write("uint GetOffset(uint attr_offset) {{\n"
+                "  return gl_VertexID * vertex_stride + attr_offset;\n"
+                "}}\n");
+    }
```

Comment:
`gl_VertexID` gets translated by spirv-cross to `SV_VertexID`.  `SV_VertexID` starts at zero regardless of what the base vertex was, while OpenGL's `gl_VertexID` (as well as Vulkan's and Metal's equivalents) start at the base vertex.

Do you have any recommendations for things to put into the comment to make it more clear?

Reply:
I wasn't aware of this base index/base vertex behavior beforehand (and I still don't entirely know where the numbers come from, as the `IDX` column in renderdoc isn't continuous between drawcalls; if it's just a buffer that new vertices are put at the end of, I'm not entirely sure how that handles differences in vertex sizes between the drawcalls, since I don't know how they would be aligned (e.g. if you have 9 vertices that are 3 bytes each (so 27 bytes total) and then a 4-byte vertex, does that get located at a 27-byte offset, or 32 bytes?)).

As for the comment itself, something like this would work (although it's a bit wordy):

> Write a function to get an offset into vertex_buffer corresponding to this vertex plus the attribute offset. This must be done differently for D3D compared to OpenGL/Vulkan/Metal, as on OpenGL etc., gl_VertexID starts counting at the base vertex specified in glDrawElementsBaseVertex, while on D3D, SV_VertexID (which spirv-cross translates gl_VertexID into) starts counting at 0 regardless of the BaseVertexLocation value passed to DrawIndexed.

(You could also link to https://registry.khronos.org/OpenGL-Refpages/gl4/html/gl_VertexID.xhtml and https://docs.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-input-assembler-stage-using#vertexid but that's probably excessive)

One other thought: it might be better to have `GetOffset()` return just the offset of the current vertex (e.g. call it `GetVertexOffset()`, and then add the attribute offset outside of it. It would be used something like this:

```glsl
float3 load_input_float3_float3(uint attr_offset) {
  uint vtx_offset = GetVertexOffset();
  return float3(uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 0]),
                uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 1]),
                uintBitsToFloat(vertex_buffer[vtx_offset + attr_offset + 2]));
}
```

Reply:
> e.g. if you have 9 vertices that are 3 bytes each (so 27 bytes total) and then a 4-byte vertex, does that get located at a 27-byte offset, or 32 bytes?)

Vulkan will align each allocation to its vertex stride: https://github.com/dolphin-emu/dolphin/blob/3f794bf87e4d90193b476cf93247b0e26aedd6a9/Source/Core/VideoBackends/Vulkan/VKStreamBuffer.cpp#L152
Metal does the same.  I haven't checked the others.

> it might be better to have GetOffset() return just the offset of the current vertex (e.g. call it GetVertexOffset(), and then add the attribute offset outside of it. It would be used something like this:

If we go forward with #10890, I'll be changing the signature to
```glsl
float3 load_input_float3_float3(uint vtx_offset, uint attr_offset) {
  uint offset = vtx_offset + attr_offset;
  return float3(uintBitsToFloat(vertex_buffer[offset + 0]),
                uintBitsToFloat(vertex_buffer[offset + 1]),
                uintBitsToFloat(vertex_buffer[offset + 2]));
}
```
to allow fetching nearby vertices' data.  Would you like me to just do that here, or stick with the separate `GetVertexOffset()` function and change it in #10890?

Reply:
That signature looks fine to me; please do include that in this PR.

> Vulkan will align each allocation to its vertex stride:

Alright, thanks. That makes sense; I just didn't find that code.

Reply:
Okay.  I updated it and pushed it to the end for you to look over.  Once you have, say so and I'll squash it in (or make changes if you request that)

Reply:
Looks good to me.